cpp/include/cudf/strings/detail/copy_range.cuh

```diff
@@ -0,0 +1,217 @@
+            (size_type idx) {
+          return (idx >= target_begin && idx < target_end) ?
+            *(source_validity_begin + (idx - target_begin)) :
+            d_target.is_valid(idx);
+        },
+        stream, mr);
+    }
+    else {  // check validities for source only
+      valid_mask = cudf::experimental::detail::valid_if(
+        thrust::make_counting_iterator<size_type>(0),
+        thrust::make_counting_iterator<size_type>(target.size()),
+        [source_validity_begin, d_target, target_begin, target_end] __device__
+            (size_type idx) {
+          return (idx >= target_begin && idx < target_end) ?
+            *(source_validity_begin + (idx - target_begin)) : true;
+        },
+        stream, mr);
+    }
+
+    auto null_count = valid_mask.second;
+    rmm::device_buffer null_mask{};
+    if (target.parent().nullable() ||
+        null_count > 0) {
+      null_mask = std::move(valid_mask.first);
+    }
+
+    // build offsets column
+
+    std::unique_ptr<column> p_offsets_column{nullptr};
+    if (target.has_nulls()) {  // check validities for both source & target
+      auto string_size_begin =
+        thrust::make_transform_iterator(
+          thrust::make_counting_iterator(0),
+          [source_value_begin, source_validity_begin, d_target, target_begin,
+              target_end] __device__ (size_type idx) {
+            if (idx >= target_begin && idx < target_end) {
+              return *(source_validity_begin + (idx - target_begin)) ?
+                (*(source_value_begin + (idx - target_begin))).size_bytes() : 0;
+            }
+            else {
+              return d_target.is_valid(idx) ?
+                d_target.element<string_view>(idx).size_bytes() : 0;
+            }
+        });
+
+      p_offsets_column =
+        detail::make_offsets_child_column(
+          string_size_begin, string_size_begin + target.size(), mr, stream);
+    }
+    else if (null_count > 0) {  // check validities for source only
+      auto string_size_begin =
+        thrust::make_transform_iterator(
+          thrust::make_counting_iterator(0),
+          [source_value_begin, source_validity_begin, d_target, target_begin,
+              target_end] __device__ (size_type idx) {
+            if (idx >= target_begin && idx < target_end) {
+              return *(source_validity_begin + (idx - target_begin)) ?
+                (*(source_value_begin + (idx - target_begin))).size_bytes() : 0;
+            }
+            else {
+              return d_target.element<string_view>(idx).size_bytes();
+            }
+        });
+
+      p_offsets_column =
+        detail::make_offsets_child_column(
+          string_size_begin, string_size_begin + target.size(), mr, stream);
+    }
+    else {  // no need to check validities
+      auto string_size_begin =
+        thrust::make_transform_iterator(
+          thrust::make_counting_iterator(0),
+          [source_value_begin, source_validity_begin, d_target, target_begin,
```

Comment:
The only difference in these 3 code paths is the lambda passed to the transform iterator. Seems like there's a potential for refactoring.

Reply:
Yeah... this was annoying to me as well, and I tried a few, but couldn't find a satisfactory solution.

Declaring nvstd::function _is_valid_ and assign different lambdas  can be the best but it seems like I cannot pass an nvstd::function object initialized in host code to device code (https://devblogs.nvidia.com/new-compiler-features-cuda-8/).

Defining a function taking a _is_valid_ functor was another, but this requires jumping back and forth to different parts of the file to understand the code with saving only 10ish lines of code.

Do you have a better solution?